lib/puppet/pops/lookup/hiera_config.rb

```diff
@@ -71,25 +57,193 @@ def self.legacy_function_config(function_name)
+    FUNCTION_PROVIDERS[function_kind].new(name, parent_data_provider, function_name, options, locations)
+  end
+
+  def self.not_implemented(impl, method_name)
+    raise NotImplementedError, "The class #{impl.class.name} should have implemented the method #{method_name}()"
+  end
+end
+
+# @api private
+class HieraConfigV4 < HieraConfig
+  require 'puppet/plugins/data_providers'
+
+  include Puppet::Plugins::DataProviders
+
+  KEY_BACKEND = 'backend'.freeze
+
+  def self.config_type
+    return @@CONFIG_TYPE if class_variable_defined?(:@@CONFIG_TYPE)
     tf = Types::TypeFactory
+    nes_t = Types::PStringType::NON_EMPTY
+
+    @@CONFIG_TYPE = tf.struct({
+      KEY_VERSION => tf.range(4, 4),
+      tf.optional(KEY_DATADIR) => nes_t,
+      tf.optional(KEY_HIERARCHY) => tf.array_of(tf.struct(
+        KEY_BACKEND => nes_t,
+        KEY_NAME => nes_t,
+        tf.optional(KEY_DATADIR) => nes_t,
+        tf.optional(KEY_PATH) => nes_t,
+        tf.optional(KEY_PATHS) => tf.array_of(nes_t)
+      ))
+    })
+  end
+
+  def factory_create_data_provider(lookup_invocation, name, parent_data_provider, provider_name, datadir, original_paths)
+    service_type = Registry.hash_of_path_based_data_provider_factories
+    provider_factory = Puppet.lookup(:injector).lookup(nil, service_type, PATH_BASED_DATA_PROVIDER_FACTORIES_KEY)[provider_name]
+    raise Puppet::DataBinding::LookupError, "#{@config_path}: No data provider is registered for backend '#{provider_name}' " unless provider_factory
+
+    paths = original_paths.map { |path| interpolate(path, lookup_invocation, false) }
+    paths = provider_factory.resolve_paths(datadir, original_paths, paths, lookup_invocation)
+
+    provider_factory_version = provider_factory.respond_to?(:version) ? provider_factory.version : 1
+    if provider_factory_version == 1
+      # Version 1 is not aware of the parent provider
+      provider_factory.create(name, paths)
+    else
+      provider_factory.create(name, paths, parent_data_provider)
+    end
+  end
+
+  def create_configured_data_providers(lookup_invocation, parent_data_provider)
+    default_datadir = @config[KEY_DATADIR]
+    data_providers = {}
+
+    @config[KEY_HIERARCHY].each do |he|
+      name = he[KEY_NAME]
+      raise Puppet::DataBinding::LookupError, "#{@config_path}: Name '#{name}' defined more than once" if data_providers.include?(name)
+      original_paths = he[KEY_PATHS] || [he[KEY_PATH] || name]
+      datadir = @config_root + (he[KEY_DATADIR] || default_datadir)
+      provider_name = he[KEY_BACKEND]
+      data_providers[name] = case provider_name
+      when 'json', 'yaml'
+        create_data_provider(name, parent_data_provider, KEY_DATA_HASH, "#{provider_name}_data", {},
+          resolve_paths(datadir, original_paths, lookup_invocation, ".#{provider_name}"))
+      else
+        # TODO: Remove support for injected provider factories
```

Comment:
I'll remove that TODO. Not relevant until we decide to actually go ahead and break backward compatibility.